extensions/notebook/src/dialog/remoteBookDialog.ts

```diff
@@ -0,0 +1,343 @@
+
+			this.versionDropdown = this.view.modelBuilder.dropDown().withProperties({
+				values: [],
+				value: '',
+				editable: false,
+			}).component();
+
+			this.versionDropdown.onValueChanged(async () => await this.fillLanguageDropdown());
+
+			this.languageDropdown = this.view.modelBuilder.dropDown().withProperties({
+				values: [],
+				value: '',
+				editable: false,
+			}).component();
+
+			this.setFieldsToEmpty();
+
+			this.formModel = this.view.modelBuilder.formContainer()
+				.withFormItems([{
+					components: [
+						{
+							component: this.remoteLocationDropdown,
+							title: loc.location,
+							required: true
+						},
+						{
+							component: this.githubRepoDropdown,
+							title: loc.repoUrl,
+							required: true
+						},
+						{
+							component: this.searchButton,
+							title: ''
+						},
+						{
+							component: this.releaseDropdown,
+							title: loc.releases,
+						},
+						{
+							component: this.bookDropdown,
+							title: loc.book,
+							required: true
+						},
+						{
+							component: this.versionDropdown,
+							title: loc.version,
+							required: true
+						},
+						{
+							component: this.languageDropdown,
+							title: loc.language,
+							required: true
+						},
+					],
+					title: ''
+				}]).withLayout({ width: '100%' }).component();
+			await this.view.initializeModel(this.formModel);
+		});
+		this.dialog.okButton.enabled = false;
+		this.dialog.okButton.label = loc.add;
+		this.dialog.cancelButton.label = loc.close;
+		this.dialog.registerCloseValidator(async () => await this.download());
+		azdata.window.openDialog(this.dialog);
+	}
+
+	private async setFieldsToEmpty(): Promise<void> {
+		await this.bookDropdown.updateProperties({
+			values: [loc.invalidTextPlaceholder],
+			value: loc.invalidTextPlaceholder
+		});
+		await this.versionDropdown.updateProperties({
+			values: [loc.invalidTextPlaceholder],
+			value: loc.invalidTextPlaceholder
+		});
+		await this.languageDropdown.updateProperties({
+			values: [loc.invalidTextPlaceholder],
+			value: loc.invalidTextPlaceholder
+		});
+	}
+
+	private get remoteLocationValue(): string {
+		return (<azdata.CategoryValue>this.remoteLocationDropdown.value).name;
+	}
+
+	public onRemoteLocationChanged(): void {
+		if (this.controller.getReleases() !== undefined && this.remoteLocationValue === loc.onGitHub) {
+			this.releaseDropdown.enabled = true;
+		} else {
+			this.releaseDropdown.enabled = false;
+		}
+	}
+
+	public async validate(): Promise<void> {
+		try {
+			let url = utils.getDropdownValue(this.githubRepoDropdown);
+			url = url.trim().toLowerCase();
+			if (this.remoteLocationValue === loc.onGitHub && url.length > 0) {
+				//get the first group to extract /owner/repo/releases format
+				let groupsRe = url.match(this.urlGithubRE);
+				if (groupsRe !== null && groupsRe.length > 0) {
+					url = loc.apiGitHub(groupsRe[0]);
+					let releases = await this.controller.fetchGithubReleases(new URL(url));
```

Comment:
Instead of having `fetchGithubReleases` and `getReleases` I would just have a single function that handles this. Either `fetchGithubReleases` that just fetches the releases (and doesn't store them anywhere) if you don't need to use it later, or `releases` (just make it a property getter) that fetches the releases if they haven't been retrieved already - otherwise returns the stored releases.

Reply:
Is there a way to make a property getter async?

Reply:
Same syntax as a function : 

```
public async get releases(): Promise<IReleaseInfo[]> {
...
```

async is just syntactic sugar - the typescript compiler just turns that method into a method that returns a promise which wraps the internals of the function. So anything that returns a Promise can be made async since at the end it's still just returning a promise.